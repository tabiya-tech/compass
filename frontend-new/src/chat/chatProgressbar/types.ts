export enum ConversationPhase {
  INITIALIZING = "INITIALIZING",
  INTRO = "INTRO",
  COLLECT_EXPERIENCES = "COLLECT_EXPERIENCES",
  DIVE_IN = "DIVE_IN",
  ENDED = "ENDED",
  UNKNOWN = "UNKNOWN"
}

export type CurrentPhase = {
  percentage: number
  phase: ConversationPhase
}

export const defaultCurrentPhase = {
  phase: ConversationPhase.INITIALIZING,
  percentage: 0
}
