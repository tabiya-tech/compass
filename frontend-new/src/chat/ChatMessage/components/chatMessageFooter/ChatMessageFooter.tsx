import React from "react";
import { Box } from "@mui/material";

export enum ChatMessageFooterChildren {
  TIMESTAMP = "TIMESTAMP",
  REACTIONS = "REACTIONS",
}

interface ChatMessageFooterProps {
  sentAt: string;
  messageId: string;
  visibleChildren: ChatMessageFooterChildren[];
  currentReaction: string;
  notifyReactionChange: (messageId: string, reaction: string) => void;
  dataTestId: string;
}

const uniqueId = "8d4e6f2c-9a3b-4c5d-b1e7-5f9d8a2b3c4e";

export const DATA_TEST_ID = {
  CHAT_MESSAGE_FOOTER_TIMESTAMP: `chat-message-footer-timestamp-${uniqueId}`,
  CHAT_MESSAGE_FOOTER_REACTIONS: `chat-message-footer-reaction-${uniqueId}`,
};

export const ChatMessageFooter: React.FC<ChatMessageFooterProps> = ({
  sentAt,
  messageId,
  currentReaction,
  notifyReactionChange,
  ...props
}) => {
  // Implement getting duration from now
  return (
    <Box display="flex" gap={1} data-testid={props["dataTestId"]}>
      {/* Implement showing timestamp and reactions based on the visible children */}
    </Box>
  );
};

export default ChatMessageFooter;
