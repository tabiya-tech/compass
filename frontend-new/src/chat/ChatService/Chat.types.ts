export interface LLMStats {
  error: string;
  prompt_token_count: number;
  response_token_count: number;
  response_time_in_sec: number;
}

export interface AgentOutput {
  message_for_user: string;
  finished: boolean;
  agent_type: string;
  reasoning: string;
  agent_response_time_in_sec: number;
  llm_stats: LLMStats[];
}

export interface Turn {
  index: number;
  input: {
    message: string;
  };
  output: AgentOutput;
}

export interface ConversationHistory {
  turns: Turn[];
}

export interface ConversationContext {
  all_history: ConversationHistory;
  history: ConversationHistory;
  summary: string;
}

export interface LastMessage {
  message_for_user: string;
  finished: boolean;
  agent_type: string;
  reasoning: string;
  agent_response_time_in_sec: number;
  llm_stats: LLMStats[];
}

export interface RootObject {
  last: LastMessage;
  conversation_context: ConversationContext;
}

export type INewChatSpecification = {
  user_id: string;
};

export type IMessageSpecification = {
  user_id: string;
  message: string;
};
