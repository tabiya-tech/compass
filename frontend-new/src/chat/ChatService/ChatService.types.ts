// Enum for the sender
export enum ConversationMessageSender {
  USER = "USER",
  COMPASS = "COMPASS",
}

// Type for individual conversation messages
export interface ConversationMessage {
  message: string;
  sent_at: string; // ISO formatted datetime string
  sender: ConversationMessageSender; // Either 'USER' or 'COMPASS'
}

export interface ConverstaionResponse {
  messages: ConversationMessage[];
  conversation_completed: boolean;
  conversation_conducted_at: string | null; // ISO formatted datetime string
  experiences_explored: number; // a count for all the experiences explored (processed)
}
