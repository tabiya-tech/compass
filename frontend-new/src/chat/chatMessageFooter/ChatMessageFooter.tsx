import React, { ReactNode } from "react";
import { Box, Divider, useTheme } from "@mui/material";

interface MessageFooterProps {
  children: ReactNode;
}

const uniqueId = "74bf72e8-1ed8-4dcf-8d3a-262d82cdd4b2";

export const DATA_TEST_ID = {
  CHAT_MESSAGE_FOOTER: `chat-message-footer-${uniqueId}`,
  CHAT_MESSAGE_FOOTER_DIVIDER: `chat-message-footer-divider-${uniqueId}`,
};

const ChatMessageFooter: React.FC<MessageFooterProps> = ({ children }) => {
  const theme = useTheme();

  return (
    <Box data-testid={DATA_TEST_ID.CHAT_MESSAGE_FOOTER}>
      <Divider
        color={theme.palette.grey[100]}
        sx={{ marginY: theme.spacing(1) }}
        data-testid={DATA_TEST_ID.CHAT_MESSAGE_FOOTER_DIVIDER}
      />
      {children}
    </Box>
  );
};

export default ChatMessageFooter;
