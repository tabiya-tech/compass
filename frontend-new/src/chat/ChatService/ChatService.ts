import { getServiceErrorFactory } from "src/error/ServiceError/ServiceError";
import { StatusCodes } from "http-status-codes";
import { fetchWithAuth } from "src/utils/fetchWithAuth/fetchWithAuth";
import ErrorConstants from "src/error/ServiceError/ServiceError.constants";
import { getBackendUrl } from "src/envService";
import { ConverstaionResponse } from "./ChatService.types";

export default class ChatService {
  readonly chatEndpointUrl: string;
  readonly chatHistoryEndpointUrl: string;
  readonly apiServerUrl: string;
  private readonly sessionId: number;

  constructor(sessionId: number) {
    this.apiServerUrl = getBackendUrl();
    this.chatEndpointUrl = `${this.apiServerUrl}/conversation`;
    this.chatHistoryEndpointUrl = `${this.apiServerUrl}/conversation/history`;
    this.sessionId = sessionId;
  }

  /**
   * Get the singleton instance of the ChatService.
   * @returns {ChatService} The singleton instance of the ChatService.
   * @param sessionId The session ID to use for the chat service.
   */
  static getInstance(sessionId: number): ChatService {
    return new ChatService(sessionId);
  }

  public getSessionId(): number {
    return this.sessionId;
  }

  public async sendMessage(message: string): Promise<ConverstaionResponse> {
    const serviceName = "ChatService";
    const serviceFunction = "sendMessage";
    const method = "POST";
    const ChatURL = this.chatEndpointUrl;
    const errorFactory = getServiceErrorFactory(serviceName, serviceFunction, method, ChatURL);

    const response = await fetchWithAuth(ChatURL, {
      method: method,
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        session_id: this.getSessionId(),
        user_input: message,
      }),
      expectedStatusCode: StatusCodes.CREATED,
      serviceName,
      serviceFunction,
      failureMessage: `Failed to send message with session id ${this.getSessionId()}`,
      expectedContentType: "application/json",
    });

    const responseBody = await response.text();

    let messageResponse: ConverstaionResponse;
    try {
      messageResponse = JSON.parse(responseBody);
    } catch (e: any) {
      throw errorFactory(
        response.status,
        ErrorConstants.ErrorCodes.INVALID_RESPONSE_BODY,
        "Response did not contain valid JSON",
        {
          responseBody,
          error: e,
        }
      );
    }

    return messageResponse;
  }

  async clearChat(): Promise<void> {
    const serviceName = "ChatService";
    const serviceFunction = "clearChat";
    const method = "GET";
    const qualifiedURL = `${this.chatEndpointUrl}?user_input=&clear_memory=true&session_id=${this.getSessionId()}`;

    await fetchWithAuth(qualifiedURL, {
      method: method,
      headers: { "Content-Type": "application/json" },
      expectedStatusCode: StatusCodes.OK,
      serviceName,
      serviceFunction,
      failureMessage: `Failed to clear chat for session id ${this.getSessionId()}`,
      expectedContentType: "application/json",
    });
  }

  public async getChatHistory(): Promise<ConverstaionResponse> {
    const serviceName = "ChatService";
    const serviceFunction = "getChatHistory";
    const method = "GET";
    const qualifiedURL = `${this.chatHistoryEndpointUrl}?session_id=${this.getSessionId()}`;

    const response = await fetchWithAuth(qualifiedURL, {
      method: method,
      headers: { "Content-Type": "application/json" },
      expectedStatusCode: StatusCodes.OK,
      serviceName,
      serviceFunction,
      failureMessage: `Failed to get chat history for session id ${this.getSessionId()}`,
      expectedContentType: "application/json",
    });

    const responseBody = await response.text();

    let chatHistory: ConverstaionResponse;
    try {
      chatHistory = JSON.parse(responseBody);
    } catch (e: any) {
      const errorFactory = getServiceErrorFactory(serviceName, serviceFunction, method, qualifiedURL);
      throw errorFactory(
        response.status,
        ErrorConstants.ErrorCodes.INVALID_RESPONSE_BODY,
        "Response did not contain valid JSON",
        {
          responseBody,
          error: e,
        }
      );
    }

    return chatHistory;
  }
}
