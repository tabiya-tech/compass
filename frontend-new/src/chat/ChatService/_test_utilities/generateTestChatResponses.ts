import { ConversationMessage, ConversationMessageSender } from "src/chat/ChatService/ChatService.types";

export const generateTestChatResponses = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};

export const generateTestHistory = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "My name is John.",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.USER,
    },
    {
      message: "Nice to meet you, John. How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};
