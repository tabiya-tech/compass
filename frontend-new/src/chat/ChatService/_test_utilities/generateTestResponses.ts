import { LastMessage, RootObject } from "src/chat/ChatService/ChatService.types";

export const generateRootObjectResponse = (): RootObject => ({
  last: {
    message_for_user: "Hello",
    finished: false,
    agent_type: "WelcomeAgent",
    reasoning: "Initial greeting",
    agent_response_time_in_sec: 1.0,
    llm_stats: [
      {
        error: "",
        prompt_token_count: 50,
        response_token_count: 10,
        response_time_in_sec: 1.0,
      },
    ],
  },
  conversation_context: {
    all_history: {
      turns: [
        {
          index: 1,
          input: {
            message: "Hello",
          },
          output: {
            message_for_user: "Hi there!",
            finished: false,
            agent_type: "WelcomeAgent",
            reasoning: "Greeting response",
            agent_response_time_in_sec: 0.5,
            llm_stats: [
              {
                error: "",
                prompt_token_count: 30,
                response_token_count: 10,
                response_time_in_sec: 0.5,
              },
            ],
          },
        },
      ],
    },
    history: {
      turns: [],
    },
    summary: "",
  },
});

export const generateLastMessageResponse = (): LastMessage => ({
  message_for_user: "Hi there!",
  finished: false,
  agent_type: "WelcomeAgent",
  reasoning: "Greeting response",
  agent_response_time_in_sec: 1.0,
  llm_stats: [
    {
      error: "",
      prompt_token_count: 30,
      response_token_count: 10,
      response_time_in_sec: 0.5,
    },
  ],
});
