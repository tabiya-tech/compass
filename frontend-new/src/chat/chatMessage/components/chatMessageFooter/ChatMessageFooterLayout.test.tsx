// mute the console
import "src/_test_utilities/consoleMock";

import ChatMessageFooter, { DATA_TEST_ID } from "src/chat/chatMessage/components/chatMessageFooter/ChatMessageFooterLayout";
import { render, screen } from "src/_test_utilities/test-utils";
import { Box } from "@mui/material";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";

describe("render tests", () => {
  // GIVEN some test children components
  const Child = (id: string) => <Box data-testid={id} key={id}>Child</Box>;

  test.each([
    {
      name: "without children",
      children: [],
      expectedTestIds: []
    },
    {
      name: "with a single child",
      children: [Child("foo-child-1")],
      expectedTestIds: ["foo-child-1"]
    },
    {
      name: "with multiple children",
      children: [Child("foo-child-1"), Child("foo-child-2")],
      expectedTestIds: ["foo-child-1", "foo-child-2"]
    }
  ])("should render a compass chat message footer %s", ({ children, expectedTestIds }) => {
    // WHEN the component is rendered
    render(<ChatMessageFooter sender={ConversationMessageSender.COMPASS}>{children}</ChatMessageFooter>);

    // THEN expect the container to be in the document
    const container = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FOOTER_LAYOUT_CONTAINER);
    expect(container).toBeInTheDocument();

    // AND expect all children to be rendered if any
    expectedTestIds.forEach(testId => {
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    // AND expect no errors or warnings
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND expect the component to match snapshot
    expect(container).toMatchSnapshot();
  });

  test.each([
    {
      name: "without children",
      children: [],
      expectedTestIds: []
    },
    {
      name: "with a single child",
      children: [Child("foo-child-1")],
      expectedTestIds: ["foo-child-1"]
    },
    {
      name: "with multiple children",
      children: [Child("foo-child-1"), Child("foo-child-2")],
      expectedTestIds: ["foo-child-1", "foo-child-2"]
    }
  ])("should render a user chat message footer %s", ({ children, expectedTestIds }) => {
    // WHEN the component is rendered
    render(<ChatMessageFooter sender={ConversationMessageSender.USER}>{children}</ChatMessageFooter>);

    // THEN expect the container to be in the document
    const container = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FOOTER_LAYOUT_CONTAINER);
    expect(container).toBeInTheDocument();

    // AND expect all children to be rendered if any
    expectedTestIds.forEach(testId => {
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    // AND expect no errors or warnings
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND expect the component to match snapshot
    expect(container).toMatchSnapshot();
  });
});