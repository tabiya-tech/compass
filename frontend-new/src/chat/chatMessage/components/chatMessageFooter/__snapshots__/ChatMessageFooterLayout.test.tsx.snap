// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render tests should render a compass chat message footer {
  name: 'with a single child',
  children: [Array],
  expectedTestIds: [Array]
} 1`] = `
<div
  class="MuiBox-root css-1qcd69q"
  data-testid="chat-message-footer-layout-container-7772f20a-9d0c-4072-b24f-97eca2f43d7b"
>
  <div
    class="MuiBox-root css-0"
    data-testid="foo-child-1"
  >
    Child
  </div>
</div>
`;

exports[`render tests should render a compass chat message footer {
  name: 'with multiple children',
  children: [Array],
  expectedTestIds: [Array]
} 1`] = `
<div
  class="MuiBox-root css-1qcd69q"
  data-testid="chat-message-footer-layout-container-7772f20a-9d0c-4072-b24f-97eca2f43d7b"
>
  <div
    class="MuiBox-root css-0"
    data-testid="foo-child-1"
  >
    Child
  </div>
  <div
    class="MuiBox-root css-0"
    data-testid="foo-child-2"
  >
    Child
  </div>
</div>
`;

exports[`render tests should render a compass chat message footer { name: 'without children', children: [], expectedTestIds: [] } 1`] = `
<div
  class="MuiBox-root css-1qcd69q"
  data-testid="chat-message-footer-layout-container-7772f20a-9d0c-4072-b24f-97eca2f43d7b"
/>
`;

exports[`render tests should render a user chat message footer {
  name: 'with a single child',
  children: [Array],
  expectedTestIds: [Array]
} 1`] = `
<div
  class="MuiBox-root css-ltxb9b"
  data-testid="chat-message-footer-layout-container-7772f20a-9d0c-4072-b24f-97eca2f43d7b"
>
  <div
    class="MuiBox-root css-0"
    data-testid="foo-child-1"
  >
    Child
  </div>
</div>
`;

exports[`render tests should render a user chat message footer {
  name: 'with multiple children',
  children: [Array],
  expectedTestIds: [Array]
} 1`] = `
<div
  class="MuiBox-root css-ltxb9b"
  data-testid="chat-message-footer-layout-container-7772f20a-9d0c-4072-b24f-97eca2f43d7b"
>
  <div
    class="MuiBox-root css-0"
    data-testid="foo-child-1"
  >
    Child
  </div>
  <div
    class="MuiBox-root css-0"
    data-testid="foo-child-2"
  >
    Child
  </div>
</div>
`;

exports[`render tests should render a user chat message footer { name: 'without children', children: [], expectedTestIds: [] } 1`] = `
<div
  class="MuiBox-root css-ltxb9b"
  data-testid="chat-message-footer-layout-container-7772f20a-9d0c-4072-b24f-97eca2f43d7b"
/>
`;
